app/src/org/commcare/adapters/EntityListAdapter.java

```diff
@@ -198,18 +207,33 @@ public TreeReference getItem(int position) {
 
     @Override
     public long getItemId(int position) {
-        if (actionsCount > 0 && position >= actionsStartPosition) {
-            return SPECIAL_ACTION;
+        int type = getItemViewType(position);
+        switch (type) {
+            case ENTITY_TYPE:
+                return references.indexOf(current.get(position).getElement());
+            case ACTION_TYPE:
+                return dividerPosition + detail.getCustomActions().indexOf(getAction(position));
+            case DIVIDER_TYPE:
+                return -2;
+            default:
+                throw new RuntimeException("Invalid view type");
         }
-        return references.indexOf(current.get(position).getElement());
+    }
+
+    private Action getAction(int position) {
+        return detail.getCustomActions().get(position - (dividerPosition + 1));
```

Comment:
the extra parens in this expression are unnecessary. If you did it that way because `dividerPosition + 1` is its own logical entity, you could just make that expression its own variable